AX: Add remote element token information to debug snapshot

hoffmanjoshua opened the pull request:
#### fc2377eb1a1bd58feeb5de992390cc63f684cc9b
<pre>
AX: Add remote element token information to debug snapshot
<a href="https://bugs.webkit.org/show_bug.cgi?id=296155">https://bugs.webkit.org/show_bug.cgi?id=296155</a>
<a href="https://rdar.apple.com/156104885">rdar://156104885</a>

Reviewed by Tyler Wilcock.

This patch adds more information about the remote elements connecting
the UI process and the web process of the main frame.

Four new pieces of information are captured and output in the snapshot:
- The token hash that the UI process generates
- The token hash that the Web Process generates
- The current remote element&apos;s token hash in the UI process
- The current remote element&apos;s token hash in the web process

By comparing these numbers, we can look for any mismatch that might have
happened.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::isAppleInternalInstall):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/C/mac/WKPagePrivateMac.mm:
(WKPageGetAccessibilityWebProcessDebugInfo):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView accessibilityRemoteChildTokenHash]):
(-[WKWebView accessibilityUIProcessLocalTokenHash]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getAccessibilityWebProcessDebugInfo):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setAccessibilityWebProcessToken):
(WebKit::WebViewImpl::accessibilityRemoteChildTokenHash):
(WebKit::WebViewImpl::accessibilityUIProcessLocalTokenHash):
(WebKit::WebViewImpl::updateRemoteAccessibilityRegistration):
(WebKit::WebViewImpl::accessibilityRegisterUIProcessTokens):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getAccessibilityWebProcessDebugInfo):
(WebKit::WebPage::clearAccessibilityIsolatedTree):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getAccessibilityWebProcessDebugInfo): Deleted.
(WebKit::WebPage::clearAccessibilityIsolatedTree): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.h:
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase setRemoteParent:token:]):
(-[WKAccessibilityWebPageObjectBase remoteTokenHash]):
(-[WKAccessibilityWebPageObjectBase setRemoteParent:]): Deleted.
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::registerRemoteFrameAccessibilityTokens):
(WebKit::WebPage::registerUIProcessAccessibilityTokens):

Canonical link: <a href="https://commits.webkit.org/297648@main">https://commits.webkit.org/297648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8530eb177fcfd3d6fbf01c7d7c275a783815c15

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112293 "Passed style check") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32024 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22502 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118371 "Built successfully") | [💥 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62657 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114255 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32677 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40588 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85307 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/62657 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115240 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26071 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101030 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65737 "Passed tests") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25376 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19166 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62217 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95456 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19245 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121697 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/110907 "Build is in progress. Recent messages:") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39367 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29296 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94115 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/111788 "Build is in progress. Recent messages:") | [⏳ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-Build-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97272 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93939 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24042 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39178 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16974 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35428 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39255 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44743 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38890 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42227 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40633 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -509,7 +509,9 @@ messages -> WebPage WantsAsyncDispatchMessage {
     ResolveAccessibilityHitTestForTesting(WebCore::FrameIdentifier frameID, WebCore::IntPoint point) -> (String result)
     EnableAccessibility()
 
+#if PLATFORM(COCOA)
```
I wonder if we should limit this entire feature to `PLATFORM(MAC)` instead since we have no current plans to expose this in a way that's useful to us on iOS 🤔  That would secure this even more. What are your thoughts?

hoffmanjoshua replied:
Ah good idea, we can always expand this down the road.

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
```diff
@@ -227,6 +227,19 @@ - (void)setRemoteParent:(id)parent
+{
+#if ENABLE(ACCESSIBILITY_ISOLATED_TREE)
+    Locker lock { m_parentLock };
+#endif // ENABLE(ACCESSIBILITY_ISOLATED_TREE)
```
It's kind of confusing to use m_parentLock here — the connection between the remote token and parent isn't immediately clear unless you're knowledgeable about how the remote-token pairing system works. I wonder if it would be more logical to just add another parameter to `setRemoteParent` so the connection is clear? That also prevents us from taking the lock a second time when we could've just taken it once.

hoffmanjoshua replied:
This is a good idea, I can update this to share `setRemoteParent`

minorninth commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -491,6 +493,9 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
     static void enableAccessibility();
     static void disableAccessibility();
+#if PLATFORM(COCOA)
+    WEBCORE_EXPORT static bool isInternalInstall();
```
How about isAppleInternalInstall (to make it clear for, e.g. GTK)?

minorninth commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1772,7 +1773,9 @@ void addResult(HashMap<WebCore::FrameIdentifier, Ref<WebCore::LegacyWebArchive>>
         @"axLiveTree": result.liveTree.createNSString().get(),
-        @"axIsolatedTree": result.isolatedTree.createNSString().get()
+        @"axIsolatedTree": result.isolatedTree.createNSString().get(),
+        @"axRemoteTokenHash": [NSNumber numberWithUnsignedInteger:result.remoteTokenHash],
```
Both tokens are "remote" depending on your perspective. What if we call this one axUIProcessGeneratedTokenHash?

Or alternatively, if we're going to display all four hashes, maybe we want:

axUIProcessLocalHash
axUIProcessRemoteHash
axWebProcessLocalHash
axWebProcessRemoteHash

where we expect axUIProcessLocalHash == axWebProcessRemoteHash, etc.

hoffmanjoshua replied:
I like that naming, I updated these two keys in the latest patch to follow that pattern, as well as the WKPagePrivate methods that we use on the UI process side to make it more clear.

hoffmanjoshua committed:
AX: Add remote element token information to debug snapshot
https://bugs.webkit.org/show_bug.cgi?id=296155
rdar://156104885

Reviewed by Tyler Wilcock.

This patch adds more information about the remote elements connecting
the UI process and the web process of the main frame.

Four new pieces of information are captured and output in the snapshot:
- The token hash that the UI process generates
- The token hash that the Web Process generates
- The current remote element's token hash in the UI process
- The current remote element's token hash in the web process

By comparing these numbers, we can look for any mismatch that might have
happened.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::isAppleInternalInstall):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/C/mac/WKPagePrivateMac.mm:
(WKPageGetAccessibilityWebProcessDebugInfo):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView accessibilityRemoteChildTokenHash]):
(-[WKWebView accessibilityUIProcessLocalTokenHash]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getAccessibilityWebProcessDebugInfo):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setAccessibilityWebProcessToken):
(WebKit::WebViewImpl::accessibilityRemoteChildTokenHash):
(WebKit::WebViewImpl::accessibilityUIProcessLocalTokenHash):
(WebKit::WebViewImpl::updateRemoteAccessibilityRegistration):
(WebKit::WebViewImpl::accessibilityRegisterUIProcessTokens):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getAccessibilityWebProcessDebugInfo):
(WebKit::WebPage::clearAccessibilityIsolatedTree):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getAccessibilityWebProcessDebugInfo): Deleted.
(WebKit::WebPage::clearAccessibilityIsolatedTree): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.h:
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase setRemoteParent:token:]):
(-[WKAccessibilityWebPageObjectBase remoteTokenHash]):
(-[WKAccessibilityWebPageObjectBase setRemoteParent:]): Deleted.
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::registerRemoteFrameAccessibilityTokens):
(WebKit::WebPage::registerUIProcessAccessibilityTokens):

Canonical link: https://commits.webkit.org/297648@main
